File: src/Dome.Server/Dome.Business/Stacks/StackProjectName.cs
using System.Text.RegularExpressions;

namespace Dome.Business.Stacks;

internal static partial class StackProjectName
{
    public const int MaxLength = 128;

    [GeneratedRegex("^[a-z0-9][a-z0-9_-]*$", RegexOptions.CultureInvariant)]
    private static partial Regex Pattern();

    public static bool TryValidate(string? projectName, out string normalized, out string? error)
    {
        normalized = string.Empty;

        if (string.IsNullOrWhiteSpace(projectName))
        {
            error = "Project name is required.";
            return false;
        }

        var trimmed = projectName.Trim();
        if (trimmed.Length > MaxLength)
        {
            error = "Project name must be 128 characters or fewer.";
            return false;
        }

        if (!Pattern().IsMatch(trimmed))
        {
            error =
                "Project name must be a valid Compose project name: lowercase letters, digits, hyphens, and underscores, starting with a letter or digit.";
            return false;
        }

        normalized = trimmed;
        error = null;
        return true;
    }
}

File: src/Dome.Client/src/lib/stackProjectName.ts
export const STACK_PROJECT_NAME_MAX_LENGTH = 128

const PROJECT_NAME_PATTERN = /^[a-z0-9][a-z0-9_-]*$/

export function validateStackProjectName(projectName: string): string | null {
  const trimmed = projectName.trim()
  if (trimmed.length === 0) {
    return 'Project name is required.'
  }

  if (trimmed.length > STACK_PROJECT_NAME_MAX_LENGTH) {
    return 'Project name must be 128 characters or fewer.'
  }

  if (!PROJECT_NAME_PATTERN.test(trimmed)) {
    return 'Use a Compose project name: lowercase letters, digits, hyphens, and underscores, starting with a letter or digit.'
  }

  return null
}
